src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2311 @@
+        else:
+            self.register_parameter("scalar", None)
+
+    def forward(self, input):
+        mean_square = torch.mean(torch.square(input), dim=self.dim, keepdim=True)
+        if self.scalar is not None:
+            input = self.scalar * input
+
+        output = input * torch.rsqrt(mean_square + self.eps)
+        return output
+
+
+class RMSNorm(nn.Module):
+    def __init__(self, number_features, eps=1e-6, affine=True):
+        super().__init__()
+        self.num_features = number_features
+        self.eps = eps
+        self.affine = affine
+        if affine:
+            self.weight = nn.Parameter(torch.Tensor(self.num_features))
+        else:
+            self.register_parameter("weight", None)
+
+    def forward(self, input):
+        mean_square = torch.mean(torch.square(input), dim=-1, keepdim=True)
+        if self.weight is not None:
+            input = input * self.weight
+
+        input * torch.rsqrt(mean_square + self.eps)
+        return input
+
+
+class MegaSequenceNorm(nn.Module):
+    def __init__(self, norm_type, embedding_dim, eps=1e-5, affine=True, export=False):
+        super().__init__()
+        if norm_type == "layernorm":
+            self.norm = nn.LayerNorm(embedding_dim, eps, elementwise_affine=affine)
+        elif norm_type == "scalenorm":
+            self.norm = ScaleNorm(dim=-1, eps=eps, affine=affine)
+        elif norm_type == "rmsnorm":
+            self.norm = RMSNorm(embedding_dim, eps=eps, affine=affine)
+        elif norm_type == "batchnorm":
+            self.norm = nn.BatchNorm1d(embedding_dim, eps=eps, affine=affine)
+        elif norm_type == "syncbatchnorm":
+            self.norm = nn.SyncBatchNorm(embedding_dim, eps=eps, affine=affine)
+        else:
+            raise ValueError("Unknown norm type: {}".format(norm_type))
+
+    def normalize(self, x):
+        if isinstance(self.norm, nn.modules.batchnorm._BatchNorm):
+            if x.dim() != 3:
+                raise ValueError("BatchNorm inputs must be exactly 3-dimensional")
+            x = x.permute(1, 2, 0)
+            x = self.norm(x)
+            return x.permute(2, 0, 1)
+        else:
+            return self.norm(x)
+
+    def forward(self, input):
+        return self.normalize(input)
+
+
+# add this layernorm class to ALL_LAYERNORM_LAYERS
+ALL_LAYERNORM_LAYERS.append(MegaSequenceNorm)
+
+
+# Dropout: standard dropout + feature dropout
+# copied from Mega repo, but changed name from Fairseq->Mega,
+# modified variable names, and removed unused items:
+# - `make_generation_fast_` method
+# - `apply_during_inference` attribute
+# - `module_name` arg
+class MegaDropout(nn.Module):
+    def __init__(self, p):
+        super().__init__()
+        self.p = p
+
+    def forward(self, input, batch_first: bool = False, inplace: bool = False):
+        if self.training:
+            return F.dropout(input, p=self.p, training=True, inplace=inplace)
+        else:
+            return input
```

Comment:
when the model is not training, dropout is not applied no? We can probably remove this entirely

Reply:
Done and coming in next commit